File: backend/domain/schema/schema.go
// SPDX-FileCopyrightText: 2025 Mercedes-Benz Group AG and Mercedes-Benz AG
//
// SPDX-License-Identifier: Apache-2.0

package schema

import (
	"time"

	"mercedes-benz.ghe.com/foss/disuko/domain"
)

type SpdxSchema struct {
	domain.RootEntity `bson:",inline"`
	Name              string     `json:"name"`
	Version           string     `json:"version"`
	Description       string     `json:"description"`
	Type              SchemaType `json:"type"`
	Active            bool       `json:"active"`
	Content           string     `json:"content,omitempty"`
	Label             string     `json:"label,omitempty"` // key of the label
}

type SchemaType int

const (
	JSON = iota
	OCTET
)

func (schema *SpdxSchema) ActivateSchema() {
	if !schema.Active {
		schema.Updated = time.Now()
		schema.Active = true
	}
}

func (schema *SpdxSchema) DeactivateSchema() {
	if schema.Active {
		schema.Updated = time.Now()
		schema.Active = false
	}
}

func (schema *SpdxSchema) MatchesProjectLabel(label string) bool {
	return schema.Label == label
}

func (schema *SpdxSchema) ContentTypeAsString() string {
	if schema.Type == JSON {
		return "application/json"
	} else {
		return "application/octet-stream"
	}
}

func (schema *SpdxSchema) ToDto() *SpdxSchemaDto {
	return ToDto(schema)
}

func ToDto(schema *SpdxSchema) *SpdxSchemaDto {
	if schema == nil {
		return nil
	}
	dto := &SpdxSchemaDto{
		Name:        schema.Name,
		Version:     schema.Version,
		Description: schema.Description,
		Type:        schema.Type,
		Active:      schema.Active,
		Content:     schema.Content,
		Label:       schema.Label,
	}
	domain.SetBaseValues(schema, dto)
	return dto
}

func ToSpdxSchemaDtoList(source []*SpdxSchema) []*SpdxSchemaDto {
	return domain.MapTo(source, ToDto)
}

File: frontend/libs/portal/model/Schema.ts
// SPDX-FileCopyrightText: 2025 Mercedes-Benz Group AG and Mercedes-Benz AG
//
// SPDX-License-Identifier: Apache-2.0

export interface ISchema {
  _key: string;
  type: number;
  content: string;
  description: string;
  created: number;
  active: boolean;
  label: string;
}

class SchemaDTO implements ISchema {
  public _key: string;
  public type: number;
  public name: string;
  public description: string;
  public content: string;
  public version: string;
  public active: boolean;
  public created: number;
  public updated: number;
  public label: string;

  public constructor() {
    this._key = '';
    this.type = 1;
    this.description = '';
    this.content = '';
    this.name = '';
    this.version = '';
    this.active = false;
    this.created = new Date().getTime();
    this.updated = new Date().getTime();
    this.label = '';
  }
}

export default class Schema extends SchemaDTO {
  constructor(dto: SchemaDTO) {
    super();
    if (dto !== null) {
      Object.assign(this, dto);
    }
  }
}
